### PR TITLE
Feature/8119 call put

### PR DESCRIPTION
**Pull request by alibombaye**
Story [AB#8119](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/8119)
Task [AB#7915](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/7915)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
     "nuke": "npm run clean && rm -rf node_modules/",
     "start:dev": "cross-env babel-node ./app/server",
     "start:redis": "docker-compose -f \"docker-compose-redis.yml\" up -d",
-    "test:ui": "cross-env NODE_ENV=test LOGGER_LEVEL=off babel-node testcafeRunner.js",
+    "test:ui": "cross-env NODE_ENV=test LOGGER_LEVEL=debug babel-node testcafeRunner.js",
```
Comment (samMcCarthyBJSS) [inline package.json:20]: left it in debug mode

File: app/pages/sections/catalogue-solutions/order-item/controller.test.js [hunk 2]
```diff
@@ -390,48 +390,105 @@ describe('catalogue-solutions order-item controller', () => {
...
+        'deliveryDate-day': '25',
+        'deliveryDate-month': '12',
+        'deliveryDate-year': '2020',
+        price: '500.49',
+        quantity: '1',
+        selectEstimationPeriod: 'month',
+      };
+
+      describe('with errors', () => {
+        it('should return error.respose if api request is unsuccessful with 400', async () => {
```
Comment (sam-roberts-bjss) [inline app/pages/sections/catalogue-solutions/order-item/controller.test.js:412]: respose?

File: app/pages/sections/catalogue-solutions/order-item/controller.js [hunk 3]
```diff
@@ -206,17 +206,76 @@ export const postSolutionOrderItem = async ({
     selectedPrice,
     formData,
   });
+
+  await postData({
+    endpoint, body, accessToken, logger,
+  });
+  logger.info(`Order item successfully created for order id: ${orderId}`);
```
Comment (sam-roberts-bjss) [inline app/pages/sections/catalogue-solutions/order-item/controller.js:213]: Might be useful to provide a little context as to what order item was created e,g. CatalogueItemName? This is obviously very minor suggestion, so doesn't need to go into this PR.

File: app/pages/sections/catalogue-solutions/order-item/controller.js [hunk 4]
```diff
@@ -206,17 +206,76 @@ export const postSolutionOrderItem = async ({
...
+        orderItemId,
+        formData,
+      });
+    return response;
   } catch (err) {
     if (err.response.status === 400 && err.response.data && err.response.data.errors) {
       return err.response.data;
     }
-    logger.error(`Error creating order item for order id: ${orderId}`);
+    logger.error(`Error saving order item for order id: ${orderId}`);
```
Comment (sam-roberts-bjss) [inline app/pages/sections/catalogue-solutions/order-item/controller.js:278]: Which order item had an error saving? Do you know at this point?

File: app/pages/sections/catalogue-solutions/order-item/flat/patient/ui-tests/withoutSavedData.ui.test.js [hunk 5]
```diff
@@ -80,14 +57,13 @@ const mocks = () => {
...
-    await serviceRecipientNameState();
+    await setState('fakeToken', authTokenInSession);
+    await setState('selectedRecipientId', selectedRecipientIdInSession);
+    await setState('selectedRecipientName', selectedRecipientNameInSession);
+    await setState('selectedSolutionId', solutionIdInSession);
+    await setState('selectedPriceId', selectedPriceIdInSession);
     if (postRoute) {
-      await solutionNameState();
-      await selectedPriceState(selectedPrice);
+      await setState('orderItemPageData', orderItemPageDataInSession);
```
Comment (sam-roberts-bjss) [inline app/pages/sections/catalogue-solutions/order-item/flat/patient/ui-tests/withoutSavedData.ui.test.js:66]: This does neaten it up 👍  Question did come to mind though, do we need anywhere that clears this information. As a user I create a new catalogue solution order item. The go back to the "newsolution" address by hand, will it bring up the data in session? Obviously edge case, so just something to consider if not already.

File: app/pages/sections/catalogue-solutions/order-item/routesHelper.test.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
+import { getPageData } from './routesHelper';
+import * as controller from './controller';
+
+const req = {};
+const fakeSessionManager = {};
+
+describe('getPageData', () => {
+  describe('when new order item', () => {
+    it('should get the selectedSolutionId from session and return this as solutionId', async () => {
+      fakeSessionManager.getFromSession = () => 'some-selected-solution-id';
```
Comment (sam-roberts-bjss) [inline app/pages/sections/catalogue-solutions/order-item/routesHelper.test.js:10]: Yes, I like that you can mock the session manager now 💯

File: app/pages/sections/catalogue-solutions/routes.test.js [hunk 7]
```diff
@@ -36,6 +37,8 @@ const mockSelectedSolutionIdCookie = 'selectedSolutionId=solution-1';
 const mockSelectedRecipientIdCookie = 'selectedRecipientId=recipient-1';
 const mockSelectedPriceIdCookie = 'selectedPriceId=1';
 
+const mockGetPageDataCookie = `orderItemPageData=${JSON.stringify({})}`;
```
Comment (sam-roberts-bjss) [inline app/pages/sections/catalogue-solutions/routes.test.js:40]: Is the JSON.stringify needed here? Could the value just be set to '{}'?

File: app/pages/sections/catalogue-solutions/order-item/controller.test.js (hunk 2 above)
Comment (alibombaye) [reply app/pages/sections/catalogue-solutions/order-item/controller.test.js:412]: yeah that should be response :joy:

File: app/pages/sections/catalogue-solutions/order-item/controller.js (hunk 4 above)
Comment (alibombaye) [reply app/pages/sections/catalogue-solutions/order-item/controller.js:278]: we do know the orderItemId if it's the put. But if it's a post we don't know the orderItemId. But I'll add the solutionName and serviceRecipient that errored

File: app/pages/sections/catalogue-solutions/order-item/flat/patient/ui-tests/withoutSavedData.ui.test.js (hunk 5 above)
Comment (alibombaye) [reply app/pages/sections/catalogue-solutions/order-item/flat/patient/ui-tests/withoutSavedData.ui.test.js:66]: That is a good point and something we haven't really considered just yet. We do need to figure out an elegant way of clearing the relevant keys from session at the right time.

File: app/pages/sections/catalogue-solutions/routes.test.js (hunk 7 above)
Comment (alibombaye) [reply app/pages/sections/catalogue-solutions/routes.test.js:40]: good point. I don't think it is needed
